support/metrics/metricsTracker.go

```diff
@@ -0,0 +1,165 @@
+package metrics
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/stellar/kelp/support/logger"
+
+	"github.com/google/uuid"
+)
+
+var (
+	amplitudeAPIURL string = "https://api2.amplitude.com/2/httpapi"
+	amplitudeAPIKey string = os.Getenv("AMPLITUDE_API_KEY")
+)
+
+// Tracker wraps the properties for Amplitude events,
+// and can be used to directly send events to the
+// Amplitude HTTP API.
+type Tracker struct {
+	client *http.Client
+	props  props
+	start  time.Time
+}
+
+type event struct {
+	UserID    string      `json:"user_id"`
+	DeviceID  string      `json:"device_id"`
+	EventType string      `json:"event_type"`
+	Props     interface{} `json:"event_properties"`
+}
+
+// props holds the properties that we need for all Amplitude events.
+// This lives on the `Tracker` struct, although
+// TODO: Add geodata.
+// TODO: Add cloud server information.
+type props struct {
+	BuildVersion string    `json:"build_version"`
+	Os           string    `json:"os"`
+	Gui          bool      `json:"gui"`
+	Strategy     string    `json:"strategy"`
+	UpdateTime   float64   `json:"update_time"`
+	Exchange     string    `json:"exchange"`
+	Pair         string    `json:"pair"`
+	SessionID    uuid.UUID `json:"session_id"`
+}
+
+// updateProps holds the properties for the update Amplitude event.
+type updateProps struct {
+	props
+	SecondsSinceStart float64 `json:"seconds_since_start"`
+}
+
+// deleteProps holds the properties for the delete Amplitude event.
+type deleteProps struct {
+	props
+	SecondsSinceStart float64 `json:"seconds_since_start"`
+	Exit              bool    `json:"exit"`
+	StackTrace        string  `json:"stack_trace"`
+}
+
+// MakeMetricsTracker is a factory method to create a `metrics.Tracker`.
+func MakeMetricsTracker(
+	version string,
+	os string,
+	gui bool,
+	strategy string,
+	updateTime float64,
+	exchange string,
+	pair string,
+) *Tracker {
+	client := &http.Client{}
+	sessionID := mustSessionID()
+	props := props{
+		BuildVersion: version,
+		Os:           os,
+		Gui:          gui,
+		Strategy:     strategy,
+		UpdateTime:   updateTime,
+		Exchange:     exchange,
+		Pair:         pair,
+		SessionID:    sessionID,
+	}
+
+	t := Tracker{
+		client: client,
+		props:  props,
+		start:  time.Now(),
+	}
+	return &t
+}
+
+func mustSessionID() uuid.UUID {
```

Comment:
The `uuid` package includes the appropriate [function](https://pkg.go.dev/github.com/google/uuid#Must). I called it this way because no other `make*` factory method returns an error in the codebase. On a non-nil error, the `make*` functions in `cmd/trade.go` call `deleteAllOffersAndExit`, which is not exported or visible here.

As per this suggestion, I've bubbled up this error to the factory method, and in turn, called `deleteOffersAndExit` if the client creation fails. But two questions:

1. I thought we didn't want to fail if on `metricsTracker` failure. This change makes it impossible to run Kelp when compiling from source, unless either the end user provides an Amplitude API key or we distribute our API key to users.
2. A more general question: within the Kelp codebase, should factory methods return an error? I don't see this anywhere and am not sure of the general best practice in Go.

Reply:
1. yes, we need to have an alternative implementation to the metricTracker along with an interface etc., that has passthrough functions. We don't need this right now, and for now we can have a separate value passed in during construction called `enabled` which, as you might think, will not send the init and sendEvent API calls if it is disabled.

2. we have quite a few factory methods that return errors, you can find them using this regex `[Mm]ake[a-zA-Z]+\(.+\)\s+\(.*error\)`. This is not a go paradigm as much as it is a general programming pattern. Throwing/returning errors during constructions helps you to fail fast and therefore fix things that are broken a lot quicker.